components/script/dom/readablestream.rs

```diff
@@ -0,0 +1,517 @@
+
+            Ok(stream)
+        }
+    }
+
+    /// Build a stream backed by a Rust underlying source.
+    #[allow(unsafe_code)]
+    pub fn new_with_external_underlying_source(
+        global: &GlobalScope,
+        source: ExternalUnderlyingSource,
+    ) -> DomRoot<ReadableStream> {
+        let _ar = enter_realm(global);
+        let cx = global.get_cx();
+
+        let source = Rc::new(ExternalUnderlyingSourceController::new(source));
+
+        let stream = ReadableStream::new(&global, Some(source.clone()));
+
+        unsafe {
+            let mut traps = ReadableStreamUnderlyingSourceTraps {
+                requestData: Some(request_data),
+                writeIntoReadRequestBuffer: Some(write_into_read_request_buffer),
+                cancel: Some(cancel),
+                onClosed: Some(close),
+                onErrored: Some(error),
+                finalize: Some(finalize),
+            };
+
+            let js_wrapper = CreateReadableStreamUnderlyingSource(
+                &mut traps,
+                &*source as *const _ as *const c_void,
+            );
+
+            rooted!(in(*cx)
+                let js_stream = NewReadableExternalSourceStreamObject(
+                    *cx,
+                    js_wrapper,
+                    ptr::null_mut() as *mut c_void,
+                    HandleObject::null(),
+                )
+            );
+
+            stream.js_stream.set(UnwrapReadableStream(js_stream.get()));
+        }
+
+        stream
+    }
+
+    /// Get a pointer to the underlying JS object.
+    pub fn get_js_stream(&self) -> NonNull<JSObject> {
+        NonNull::new(self.js_stream.get())
+            .expect("Couldn't get a non-null pointer to JS stream object.")
+    }
+
+    #[allow(unsafe_code)]
+    pub fn enqueue_native(&self, bytes: Vec<u8>) {
+        let global = self.global();
+        let _ar = enter_realm(&*global);
+        let cx = global.get_cx();
+
+        let stream_handle = unsafe { self.js_stream.handle() };
+
+        self.external_underlying_source
+            .as_ref()
+            .expect("No external source to enqueue bytes.")
+            .enqueue_chunk(cx, stream_handle, bytes);
+    }
+
+    #[allow(unsafe_code)]
+    pub fn error_native(&self, error: Error) {
+        let global = self.global();
+        let _ar = enter_realm(&*global);
+        let cx = global.get_cx();
+
+        unsafe {
+            rooted!(in(*cx) let mut js_error = UndefinedValue());
+            error.to_jsval(*cx, &global, js_error.handle_mut());
+            ReadableStreamError(
+                *cx,
+                self.js_stream.handle(),
+                js_error.handle().into_handle(),
+            );
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn close_native(&self) {
+        let global = self.global();
+        let _ar = enter_realm(&*global);
+        let cx = global.get_cx();
+
+        let handle = unsafe { self.js_stream.handle() };
```

Comment:
Why is this unsafe?

Reply:
the call to `handle` is actually unsafe.